### PR TITLE
Adding data to minc header

### PR DESCRIPTION
**Pull request by nist-anka**
I still don't know how to get the time stamp.   I have added a variable acquisition and under it frame ID, Calibration matrix and flag telling if the frame was saved with calibration matrix applied.

### CODE REVIEW

File: IbisLib/usacquisitionobject.cpp [hunk 1]
```diff
@@ -979,6 +980,33 @@ void USAcquisitionObject::ExportTrackedVideoBuffer(QString destDir , bool masked
...
+    calMatString.append( ", ");
+    calMatString.append( QString::number(calMatrix->GetElement( 1, 2 ), 'g', 6 ) );
+    calMatString.append( ", ");
+    calMatString.append( QString::number(calMatrix->GetElement( 2, 0 ), 'g', 6 ) );
+    calMatString.append( ", ");
+    calMatString.append( QString::number(calMatrix->GetElement( 2, 1 ), 'g', 6 ) );
+    calMatString.append( ", ");
+    calMatString.append( QString::number(calMatrix->GetElement( 2, 2 ), 'g', 6 ) );
+    calMatString.append( ", ");
+    calMatString.append( QString::number(calMatrix->GetElement( 2, 3 ), 'g', 6 ) );
```
Comment (hgueziri) [inline IbisLib/usacquisitionobject.cpp:1008]: Preferably use a loop to shorten the code

**Comment by hgueziri** [comment]
The timestamp is associated with USProbeObject (inherited from TrackedSceneObject). You need to add accessors to get it while recording/exporting/importing in USAcquisitionObject specifically lines [316](https://github.com/IbisNeuronav/Ibis/blob/de5e445500a628c489bbb42e103b3b74ca58583c/IbisLib/usacquisitionobject.cpp#L316), [336](https://github.com/IbisNeuronav/Ibis/blob/de5e445500a628c489bbb42e103b3b74ca58583c/IbisLib/usacquisitionobject.cpp#L336) and [613](https://github.com/IbisNeuronav/Ibis/blob/de5e445500a628c489bbb42e103b3b74ca58583c/IbisLib/usacquisitionobject.cpp#L613). 

TrackedVideoBuffer needs to have a timestamp associated with each added frame that can be added if needed using [AddFrame()](https://github.com/IbisNeuronav/Ibis/blob/5316e1b65539e8c2942ec9c87cde3ac0df54a8a4/IbisLib/trackedvideobuffer.h#L40). Then we can add a method GetTimestamp(int frameId).

In USAcquisitionObject in method [GetItkImage()](https://github.com/IbisNeuronav/Ibis/blob/76137641b04607699d3f403e1d1badcaf93daad6/IbisLib/usacquisitionobject.cpp#L777) m_videoBuffer->GetTimestamp(frameNo) needs to be called to get the timestamp and returned in lines [1037 and 1083](https://github.com/IbisNeuronav/Ibis/blob/76137641b04607699d3f403e1d1badcaf93daad6/IbisLib/usacquisitionobject.cpp#L1037)

Hope that helps

**Comment by nist-anka** [comment]
I agree, a loop is better.

File: IbisLib/usacquisitionobject.cpp [hunk 2]
```diff
@@ -983,29 +983,15 @@ void USAcquisitionObject::ExportTrackedVideoBuffer(QString destDir , bool masked
...
-    calMatString.append( QString::number(calMatrix->GetElement( 2, 2 ), 'g', 6 ) );
-    calMatString.append( ", ");
-    calMatString.append( QString::number(calMatrix->GetElement( 2, 3 ), 'g', 6 ) );
+    for( int i = 0; i < 3; i++ )
+    {
+        for( int j = 0; j < 4; j++ )
+        {
+            calMatString.append( QString::number(calMatrix->GetElement( i, j ), 'g', 6 ) );
+            if( j < 3 )
+                calMatString.append( ", ");
```
Comment (hgueziri) [inline IbisLib/usacquisitionobject.cpp:992]: Not sure this works, the "," will not be in the right place.
I wonder if we shouldn't record the transform as in .xfm format, with a space separator and ends with ";". Something like:
Linear_Transform =
T_00 T_01 T02 T03
T_10 T_11 T12 T13
T_20 T_21 T22 T23;

File: IbisLib/usacquisitionobject.cpp [hunk 3]
```diff
@@ -1036,11 +1022,12 @@ void USAcquisitionObject::ExportTrackedVideoBuffer(QString destDir , bool masked
                 IbisItkUnsignedChar3ImageType::Pointer itkSliceImage = IbisItkUnsignedChar3ImageType::New();
                 this->GetItkImage(itkSliceImage, i,  masked,  useCalibratedTransform, relativeToID );
                 //Output acquisition properties: time stamp, calibration matrix, frame ID, flag telling idf the calibration matrix was applied
-                // TODO: Somehow obtain time stamp of the frame and output it also
+                double timestamp = m_videoBuffer->GetTimestamp( i );
                 itk::MetaDataDictionary & metaDict = itkSliceImage->GetMetaDataDictionary();
                 itk::EncapsulateMetaData< std::string >( metaDict, "acquisition:calibratioMatrix", calMatString.toUtf8().data() );
                 itk::EncapsulateMetaData< std::string >( metaDict, "acquisition:calibratioMatrixApplied", useCalibratedTransform?"true":"false" );
-                itk::EncapsulateMetaData< std::string >( metaDict, "acquisition:frameID", Number.toUtf8().data() );
+                itk::EncapsulateMetaData< std::string >( metaDict, "acquisition:timestamp", QString::number( timestamp, 'g', 6 ).toUtf8().data() );
```
Comment (hgueziri) [inline IbisLib/usacquisitionobject.cpp:1029]: should use format 'f'

File: IbisLib/usacquisitionobject.cpp [hunk 4]
```diff
@@ -1083,10 +1070,11 @@ void USAcquisitionObject::ExportTrackedVideoBuffer(QString destDir , bool masked
                 IbisRGBImageType::Pointer itkSliceImage = IbisRGBImageType::New();
                 this->GetItkRGBImage(itkSliceImage, i,  masked,  useCalibratedTransform, relativeToID );
                 //Output acquisition properties: time stamp, calibration matrix, frame ID, flag telling idf the calibration matrix was applied
-                // TODO: Somehow obtain time stamp of the frame and output it also
+                double timestamp = m_videoBuffer->GetTimestamp( i );
                 itk::MetaDataDictionary & metaDict = itkSliceImage->GetMetaDataDictionary();
                 itk::EncapsulateMetaData< std::string >( metaDict, "acquisition:calibratioMatrix", calMatString.toUtf8().data() );
                 itk::EncapsulateMetaData< std::string >( metaDict, "acquisition:calibratioMatrixApplied", useCalibratedTransform?"true":"false" );
+                itk::EncapsulateMetaData< std::string >( metaDict, "acquisition:timestamp", QString::number( timestamp, 'g', 6 ).toUtf8().data() );
```
Comment (hgueziri) [inline IbisLib/usacquisitionobject.cpp:1077]: should use format 'f'

File: IbisLib/usacquisitionobject.cpp (hunk 2 above)
Comment (nist-anka) [reply IbisLib/usacquisitionobject.cpp:992]: Single line will be ok?

**Comment by nist-anka** [comment]
Not finished yet, I have to read the timestamps fro files.

**Comment by nist-anka** [comment]
Maybe I should use 1 and 0 instead true and false for the flag?

**Comment by nist-anka** [comment]
Still not finished

**Comment by hgueziri** [comment]
It would be good to add accessors to the timestamp from a USAcquisitionObject. For now, they are accessible internally using videoBuffer (only from within the USAcquisition class).

**Comment by nist-anka** [comment]
OK, I'll do it

**Comment by hgueziri** [comment]
Works on windows. Acquisition timestamps now save and load perfectly. Thank you!
Small detail: need to add "**m_timestamps.clear();**" in [TrackedVideoBuffer::Clear()](https://github.com/IbisNeuronav/Ibis/blob/ec40f4f29d198fe28fb9dd4cabef0830ab80994e/IbisLib/trackedvideobuffer.cpp#L38) to initialize the list.

**Comment by nist-anka** [comment]
I'll do it

**Pull request merged**